src/library/scala/collection/immutable/ListSet.scala

```diff
@@ -174,9 +170,17 @@ sealed class ListSet[A] extends AbstractSet[A]
 
     /** `-` can be used to remove a single element from a set.
      */
-    override def -(e: A): ListSet[A] = if (e == head) self else {
-      val tail = self - e; new tail.Node(head)
-    }
+    override def -(e: A): ListSet[A] = removeInternal(e, this, Nil)
+
+    @tailrec private def removeInternal(k: A, cur: ListSet[A], acc: List[ListSet[A]]): ListSet[A] =
+      if (cur.isEmpty)
+        acc.last
+      else if (k == cur.head)
+        (cur.tail /: acc) {
+          case (t, h) => val tt = t; new tt.Node(h.head)
```

Comment:
according to the comment in ListMap, `tt` exists due to https://issues.scala-lang.org/browse/SI-7459, which is marked as fixed now. could you remove `tt` in both places?